src/openms_gui/source/VISUAL/DIALOGS/FLASHDeconvTabWidget.cpp

```diff
@@ -0,0 +1,436 @@
+// --------------------------------------------------------------------------
+//                   OpenMS -- Open-Source Mass Spectrometry
+// --------------------------------------------------------------------------
+// Copyright The OpenMS Team -- Eberhard Karls University Tuebingen,
+// ETH Zurich, and Freie Universitaet Berlin 2002-2020.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Jihyung Kim $
+// $Authors: Jihyung Kim $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/VISUAL/DIALOGS/FLASHDeconvTabWidget.h>
+#include <ui_FLASHDeconvTabWidget.h>
+
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/FORMAT/FileHandler.h>
+#include <OpenMS/FORMAT/ParamXMLFile.h>
+#include <OpenMS/SYSTEM/File.h>
+#include <OpenMS/VISUAL/DIALOGS/TOPPASInputFilesDialog.h>
+#include <OpenMS/VISUAL/MISC/GUIHelpers.h>
+
+#include <QtCore/QDateTime>
+#include <QtCore/QDir>
+#include <QMessageBox>
+#include <QProcess>
+#include <QProgressDialog>
+#include <QSignalBlocker>
+#include <QDesktopServices>
+
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+  namespace Internal
+  {
+
+    FLASHDeconvGUILock::FLASHDeconvGUILock(FLASHDeconvTabWidget* ftw)
+      :
+      ftw_(ftw),
+      old_(ftw->currentWidget()),
```

Comment:
is this order thread-safe?
e.g., old_(ftw->currentWidget()) get's executed before the lock is acquired?
Did not look into this in detail.

Reply:
this should not matter because: 
a) this lock should only ever be acquired by the GUI thread, not a worker thread 
b) the lock happens in the Ctor, so only one thread can get there. And the lock is specific to this object.